packages/gnarly-reducer-erc20/src/reducer.ts

```diff
@@ -0,0 +1,136 @@
+import BN = require('bn.js')
+import makeDebug = require('debug')
+const debug = makeDebug('gnarly-reducer:erc20')
+
+import {
+  addABI,
+  Block,
+  getLogs,
+  IReducer,
+  ITypeStore,
+  ReducerType,
+  toHex,
+} from '@xlnt/gnarly-core'
+
+import abi from './abi'
+
+const TRANSFER_REASON = 'ERC20_TRANSFER'
+const MINT_REASON = 'ERC20_MINT'
+const BURN_REASON = 'ERC20_BURN'
+
+const isZeroAddress = (address: string) => address === '0x0000000000000000000000000000000000000000'
+
+const makeReducer = (key: string, typeStore: ITypeStore) => (
+  tokenAddress: string,
+): IReducer => {
+  addABI(tokenAddress, abi)
+
+  interface IERC20Tracker {
+    balances: {
+      [id: string]: {
+        id: string,
+        tokenAddress: string;
+        owner: string;
+        balance: string;
+        balanceStr: string;
+      },
+    },
+  }
+
+  const makeActions = (state: IERC20Tracker, { operation, emit }) => ({
+    transfer: (from: string, to: string, rawTokenValue: string) => {
+      debug('transferring %d tokens from %s to %s', rawTokenValue, from, to)
+
+      // turn rawTokenValue into BN tokenValue
+      const tokenValue = new BN(rawTokenValue)
+
+      // hacky composite key for O(n) lookups required by JSON-Patch
+      const fromId = `${tokenAddress}-${from}`
+      const toId = `${tokenAddress}-${to}`
+
+      // create new addresses' states which are missing
+      // order dependent, operations must be done before appendTo
+      if (!state.balances[fromId]) {
+        operation(() => {
+          const balance = tokenValue.toString()
+
+          state.balances[fromId] = {
+            id: fromId,
+            tokenAddress,
+            owner: from,
+            balance,
+            balanceStr: balance,
+          }
+        })
+      }
+
+      if (!state.balances[toId]) {
+        const balance = '0'
+
+        operation(() => {
+          state.balances[toId] = {
+            id: toId,
+            tokenAddress,
+            owner: to,
+            balance,
+            balanceStr: balance,
+          }
+        })
+      }
+
+      operation(() => {
```

Comment:
nit: this doesn't actually need to be within an operation; it's implied by the `because` block that this is executed in.

Reply:
I see! functions like because / operation / emit etc, we could make some documentation on them cause I am not sure how they work tbh

Reply:
you're totally right 😁

I'm not sure if it's worth investing in a docs wiki or anything just yet, so I'll add an issue to append this information to the README